applications/FluidDynamicsApplication/custom_utilities/fluid_adjoint_test_utilities.cpp

```diff
@@ -0,0 +1,324 @@
+    const double BossakAlpha,
+    const Variable<TDataType>& rVariable,
+    const NodeType& rNode)
+{
+    return (1 - BossakAlpha) * rNode.FastGetSolutionStepValue(rVariable) +
+           BossakAlpha * rNode.FastGetSolutionStepValue(rVariable, 1);
+}
+
+
+template<>
+std::function<double&(ModelPart::NodeType&, const IndexType)> FluidAdjointTestUtilities::GetPerturbationMethod(
+    const Variable<double>& rPerturbationVariable)
+{
+    auto perturbation_method = [&rPerturbationVariable](
+                                   NodeType& rNode, const IndexType Dimension) -> double& {
+        KRATOS_TRY
+
+        KRATOS_DEBUG_ERROR_IF(Dimension != 0)
+            << "Dimension should be always 0 for scalar perturbation "
+               "variables. [ Perturbation variable = "
+            << rPerturbationVariable.Name() << ", Dimension = " << Dimension << " ].\n";
+
+        return rNode.FastGetSolutionStepValue(rPerturbationVariable);
+
+        KRATOS_CATCH("");
+    };
+
+    return perturbation_method;
+}
+
+template <>
+std::function<double&(ModelPart::NodeType&, const IndexType)> FluidAdjointTestUtilities::GetPerturbationMethod(
+    const Variable<array_1d<double, 3>>& rPerturbationVariable)
+{
+    if (rPerturbationVariable == SHAPE_SENSITIVITY) {
+        auto perturbation_method = [](NodeType& rNode, const IndexType iDim) -> double& {
+            array_1d<double, 3>& r_coordinates = rNode.Coordinates();
+            return r_coordinates[iDim];
+        };
+        return perturbation_method;
+    } else {
+        auto perturbation_method = [&rPerturbationVariable](
+                                       NodeType& rNode, const IndexType Dimension) -> double& {
+            KRATOS_TRY
+
+            KRATOS_DEBUG_ERROR_IF(Dimension > 2)
+                << "Dimension should be always less than 3 for scalar "
+                   "perturbation "
+                   "variables. [ Perturbation variable = "
+                << rPerturbationVariable.Name() << ", Dimension = " << Dimension
+                << " ].\n";
+
+            array_1d<double, 3>& r_vector =
+                rNode.FastGetSolutionStepValue(rPerturbationVariable);
+            return r_vector[Dimension];
+
+            KRATOS_CATCH("");
+        };
+        return perturbation_method;
+    }
+}
+
+template <>
+IndexType FluidAdjointTestUtilities::GetVariableDimension(
+    const Variable<double>& rVariable,
+    const ProcessInfo& rProcessInfo)
+{
+    return 1;
+}
+
+template <>
+IndexType FluidAdjointTestUtilities::GetVariableDimension(
+    const Variable<array_1d<double, 3>>& rVariable,
+    const ProcessInfo& rProcessInfo)
+{
+    return rProcessInfo[DOMAIN_SIZE];
+}
+
+template<class TContainerType>
+void FluidAdjointTestUtilities::Testing<TContainerType>::CalculateResidual(
+    Vector& residual,
+    typename TContainerType::data_type& rEntity,
+    const ProcessInfo& rProcessInfo)
+{
+    const double bossak_alpha = rProcessInfo[BOSSAK_ALPHA];
```

Comment:
Standard QSVMS supports BDF2 (just mentioning in case this could affect the tests).

Reply:
I added a comment as a reminder for future me :). Thanks